### DISCUSSION TITLE
Will Socket.SendAsync send everything in 1 call?

### DISCUSSION THREAD

**Discussion by xqrzd** [dotnet/runtime#100436] [Q&A] [answered]
Socket.SendAsync() returns an int, does this mean we have to loop in case it doesn't send everything in one call, or is it guaranteed to send the entire buffer?

Is this safe?
```csharp
ReadOnlyMemory<byte> buffer = ...
await socket.SendAsync(buffer);
```

or is this required?
```csharp
ReadOnlyMemory<byte> buffer = ...
while (buffer.Length > 0)
{
    var written = await socket.SendAsync(buffer);
    buffer = buffer[written..];
}
```

**Comment by Clockwork-Muse** [comment]
Per [the documentation](https://learn.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.sendasync?view=net-8.0#system-net-sockets-socket-sendasync(system-readonlymemory((system-byte))-system-threading-cancellationtoken)):

> Returns
> [ValueTask](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask-1?view=net-8.0)<[Int32](https://learn.microsoft.com/en-us/dotnet/api/system.int32?view=net-8.0)>
> An asynchronous task that completes with the number of bytes sent.

Yes, the loop will be required.

**Reply by xqrzd** [reply]
Thanks. I thought maybe it wouldn't be since NetworkStream is a thin wrapper on top and didn't return a byte count.

**Reply by Clockwork-Muse** [reply]
`NetworkStream` does some funny things under the covers, it doesn't actually (directly) wrap this method.  Keep in mind there's a different async API for sockets as well....

**Comment by stephentoub** [comment, marked as answer]
We should update the docs. As long as the Socket.Blocking is true, for a connection-oriented protocol (e.g. TCP), Socket.SendAsync will ensure all the data has been sent (from the perspective of the application) and you do not need a loop. This is why NetworkStream.WriteAsync doesn't have or need a loop.

On Windows, the implementation uses WSASend, which makes this guarantee. And on Unix, the implementation has its own loop to provide a similar guarantee.

**Reply by xqrzd** [reply]
Thank you!

**Reply by kerryjiang** [reply]
The doc still doesn't indicate this information:
https://learn.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.sendasync?view=net-8.0

**Reply by kasperk81** [reply]
https://learn.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.blocking?view=net-8.0#remarks says
> The [Blocking](https://learn.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.blocking?view=net-8.0) property has no effect on asynchronous methods. If you are sending and receiving data asynchronously and want to block execution, use the [ManualResetEvent](https://learn.microsoft.com/en-us/dotnet/api/system.threading.manualresetevent?view=net-8.0) class.

is this correct?

**Reply by murshex** [reply]
@stephentoub Please update docs
